File: android/app/src/main/kotlin/com/endurain/endurain/activity/ActiveActivityStore.kt
package com.endurain.endurain.activity

import android.content.Context
import android.util.AtomicFile
import java.io.File
import java.io.FileNotFoundException
import org.json.JSONObject

/**
 * App-private, file-backed store for the single active recording.
 *
 * Mirrors the Dart `FileActiveActivityStore` layout so either recorder can
 * recover state written by the other:
 *
 * ```
 * <filesDir>/activity_records/active/session.json
 * <filesDir>/activity_records/active/points.jsonl
 * ```
 *
 * On Android `path_provider`'s application-support directory resolves to
 * `context.filesDir`, so this path matches the Dart store exactly. All files
 * live in app-private storage; nothing is written to shared/public storage.
 */
class ActiveActivityStore private constructor(appContext: Context) {
    private val rootDir = File(appContext.filesDir, ROOT_DIR)
    private val activeDir = File(rootDir, ACTIVE_DIR)
    private val sessionFile = File(activeDir, SESSION_FILE)
    private val atomicSessionFile = AtomicFile(sessionFile)
    private val pointsFile = File(activeDir, POINTS_FILE)
    private val pointsLog = ActivePointsLog(pointsFile)
    private val announcementFile = File(activeDir, ANNOUNCEMENT_FILE)
    private val atomicAnnouncementFile = AtomicFile(announcementFile)
    private val lock = Any()

    fun saveSession(session: ActiveActivitySessionData) {
        synchronized(lock) {
            ensureDir()
            var output: java.io.FileOutputStream? = null
            try {
                output = atomicSessionFile.startWrite()
                output.write(session.toJson().toString().toByteArray(Charsets.UTF_8))
                output.fd.sync()
                atomicSessionFile.finishWrite(output)
            } catch (error: Exception) {
                output?.let { atomicSessionFile.failWrite(it) }
                throw error
            }
        }
    }

    fun loadSession(): ActiveActivitySessionData? {
        synchronized(lock) {
            return try {
                atomicSessionFile.openRead().bufferedReader().use { reader ->
                    ActiveActivitySessionData.fromJson(JSONObject(reader.readText()))
                } ?: pointsLog.orphanedSession()
            } catch (_: FileNotFoundException) {
                pointsLog.orphanedSession()
            } catch (_: Exception) {
                pointsLog.orphanedSession()
            }
        }
    }

    fun hasRecoverableData(): Boolean {
        synchronized(lock) {
            return activeDir.listFiles()?.isNotEmpty() == true
        }
    }

    fun appendPoints(points: List<RecordedActivityPointData>) {
        if (points.isEmpty()) {
            return
        }
        synchronized(lock) {
            ensureDir()
            pointsLog.append(points)
        }
    }

    /**
     * Reads stored points, skipping the first [sinceOffset] valid points.
     *
     * Malformed lines are ignored and do not advance the offset, matching the
     * Dart store's drain semantics.
     */
    fun readPoints(sinceOffset: Int = 0): List<RecordedActivityPointData> {
        synchronized(lock) {
            return pointsLog.read(sinceOffset)
        }
    }

    /** Counts only valid, parseable points so callers can compute offsets. */
    fun pointCount(): Int {
        synchronized(lock) {
            return pointsLog.count()
        }
    }

    /** Returns the last valid point for segment-gap recovery after restart. */
    fun lastPoint(): RecordedActivityPointData? {
        synchronized(lock) {
            return pointsLog.last()
        }
    }

    /** Removes all active-recording files. */
    fun clear() {
        synchronized(lock) {
            if (activeDir.exists()) {
                activeDir.deleteRecursively()
            }
        }
    }

    /**
     * Persists the audio-announcement config + progress for the active
     * recording. Lives inside the same active-recording directory as the
     * session/points files, so [clear] removes it too and [hasRecoverableData]
     * keeps working unchanged.
     */
    fun saveAnnouncementState(state: AnnouncementStateData) {
        synchronized(lock) {
            ensureDir()
            var output: java.io.FileOutputStream? = null
            try {
                output = atomicAnnouncementFile.startWrite()
                output.write(state.toJson().toString().toByteArray(Charsets.UTF_8))
                output.fd.sync()
                atomicAnnouncementFile.finishWrite(output)
            } catch (error: Exception) {
                output?.let { atomicAnnouncementFile.failWrite(it) }
                throw error
            }
        }
    }

    /** Returns the persisted announcement state, or `null` when absent/corrupt. */
    fun loadAnnouncementState(): AnnouncementStateData? {
        synchronized(lock) {
            return try {
                atomicAnnouncementFile.openRead().bufferedReader().use { reader ->
                    AnnouncementStateData.fromJson(JSONObject(reader.readText()))
                }
            } catch (_: FileNotFoundException) {
                null
            } catch (_: Exception) {
                null
            }
        }
    }

    private fun ensureDir() {
        if (!activeDir.exists()) {
            activeDir.mkdirs()
        }
    }

    companion object {
        const val ROOT_DIR = "activity_records"
        const val ACTIVE_DIR = "active"
        const val SESSION_FILE = "session.json"
        const val POINTS_FILE = "points.jsonl"
        const val ANNOUNCEMENT_FILE = "announcement.json"

        @Volatile
        private var instance: ActiveActivityStore? = null

        /** Process-wide singleton so the service and channel share one lock. */
        fun of(context: Context): ActiveActivityStore {
            return instance ?: synchronized(this) {
                instance ?: ActiveActivityStore(context.applicationContext)
                    .also { instance = it }
            }
        }
    }
}

File: ios/Runner/Activity/ActiveActivityStore.swift
import Foundation

/// File-backed store for the in-progress activity recording on iOS.
///
/// Mirrors the Android `ActiveActivityStore` and the Dart active store layout.
/// All files live under app-private application support storage
/// (`Library/Application Support/activity_records/active/`) which matches the
/// path returned by `path_provider`'s `getApplicationSupportDirectory()` so the
/// Dart and native sides agree on the location. Never writes to shared or
/// publicly visible storage.
final class ActiveActivityStore {
    static let shared = ActiveActivityStore()

    private let fileManager = FileManager.default
    private let queue = DispatchQueue(label: "com.endurain.activity.store")

    private init() {}

    private var activeDirectory: URL {
        let base = fileManager.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
        return base
            .appendingPathComponent("activity_records", isDirectory: true)
            .appendingPathComponent("active", isDirectory: true)
    }

    private var sessionFile: URL {
        return activeDirectory.appendingPathComponent("session.json", isDirectory: false)
    }

    private var pointsFile: URL {
        return activeDirectory.appendingPathComponent("points.jsonl", isDirectory: false)
    }

    private func ensureDirectory() throws {
        if !fileManager.fileExists(atPath: activeDirectory.path) {
            try fileManager.createDirectory(
                at: activeDirectory,
                withIntermediateDirectories: true
            )
        }
    }

    func saveSession(_ session: ActiveActivitySessionData) {
        queue.sync {
            guard let json = session.toJsonString() else { return }
            do {
                try ensureDirectory()
                try json.data(using: .utf8)?.write(to: sessionFile, options: .atomic)
            } catch {
                // Surface persistence failures via the coordinator; never log
                // the session payload itself.
                ActivityRecorderCoordinator.shared.emitFailed(
                    ActivityRecorderCoordinator.reasonPersistenceFailed
                )
            }
        }
    }

    func loadSession() -> ActiveActivitySessionData? {
        return queue.sync {
            guard
                let data = try? Data(contentsOf: sessionFile),
                let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
            else {
                return orphanedSession()
            }
            return ActiveActivitySessionData.fromJson(object) ?? orphanedSession()
        }
    }

    func hasRecoverableData() -> Bool {
        return queue.sync {
            guard let files = try? fileManager.contentsOfDirectory(
                at: activeDirectory,
                includingPropertiesForKeys: nil
            ) else {
                return false
            }
            return !files.isEmpty
        }
    }

    /// Appends already-validated points to the JSONL file. Throws on I/O
    /// failure so callers can surface a typed recorder failure.
    func appendPoints(_ points: [RecordedActivityPointData]) throws {
        if points.isEmpty {
            return
        }
        try queue.sync {
            try ensureDirectory()
            var payload = Data()
            for point in points {
                guard let line = point.toJsonLine() else { continue }
                if let lineData = (line + "\n").data(using: .utf8) {
                    payload.append(lineData)
                }
            }
            if payload.isEmpty {
                return
            }
            if !fileManager.fileExists(atPath: pointsFile.path) {
                try payload.write(to: pointsFile)
            } else {
                // Throwing variants (iOS 13.4+) rather than the ObjC-bridged
                // `seekToEndOfFile()`/`write(_:)`/`closeFile()`, which raise an
                // NSException on I/O failure. Swift cannot catch NSException, so
                // those would bypass the caller's do/catch — which exists
                // specifically to convert an append failure into a typed
                // `persistenceFailed` — and hard-crash the app mid-recording.
                let handle = try FileHandle(forWritingTo: pointsFile)
                defer { try? handle.close() }
                try handle.seekToEnd()
                try handle.write(contentsOf: payload)
            }
        }
    }

    /// Reads persisted points, skipping malformed lines, returning entries at or
    /// after `sinceOffset` (counted across valid points only).
    func readPoints(sinceOffset: Int = 0) throws -> [RecordedActivityPointData] {
        return try queue.sync {
            guard fileManager.fileExists(atPath: pointsFile.path) else {
                return []
            }
            let content = try String(contentsOf: pointsFile, encoding: .utf8)
            var result: [RecordedActivityPointData] = []
            var validIndex = 0
            for rawLine in content.split(separator: "\n", omittingEmptySubsequences: false) {
                guard let point = RecordedActivityPointData.tryParseLine(String(rawLine)) else {
                    continue
                }
                if validIndex >= sinceOffset {
                    result.append(point)
                }
                validIndex += 1
            }
            return result
        }
    }

    func pointCount() -> Int {
        return queue.sync {
            guard let content = try? String(contentsOf: pointsFile, encoding: .utf8) else {
                return 0
            }
            var count = 0
            for rawLine in content.split(separator: "\n", omittingEmptySubsequences: false) {
                if RecordedActivityPointData.tryParseLine(String(rawLine)) != nil {
                    count += 1
                }
            }
            return count
        }
    }

    func lastPoint() -> RecordedActivityPointData? {
        return queue.sync {
            guard let content = try? String(contentsOf: pointsFile, encoding: .utf8) else {
                return nil
            }
            var last: RecordedActivityPointData?
            for rawLine in content.split(separator: "\n", omittingEmptySubsequences: false) {
                if let point = RecordedActivityPointData.tryParseLine(String(rawLine)) {
                    last = point
                }
            }
            return last
        }
    }

    func clear() {
        queue.sync {
            try? fileManager.removeItem(at: activeDirectory)
        }
    }

    private func orphanedSession() -> ActiveActivitySessionData? {
        guard
            let content = try? String(contentsOf: pointsFile, encoding: .utf8)
        else {
            return nil
        }
        let points = content
            .split(separator: "\n", omittingEmptySubsequences: false)
            .compactMap { RecordedActivityPointData.tryParseLine(String($0)) }
        guard let first = points.first, let last = points.last else {
            return nil
        }
        let firstMillis = IsoTime.toEpochMillis(first.timestamp) ?? 0
        let lastMillis = IsoTime.toEpochMillis(last.timestamp) ?? firstMillis
        return ActiveActivitySessionData(
            localSessionId: "recovered_\(Int(fileManager.modificationDate(pointsFile).timeIntervalSince1970))",
            activityType: "other",
            status: ActiveActivitySessionData.statusFailed,
            startedAt: first.timestamp,
            endedAt: last.timestamp,
            elapsedDurationSeconds: max(0, Int((lastMillis - firstMillis) / 1000)),
            currentSegmentIndex: last.segmentIndex
        )
    }
}

private extension FileManager {
    func modificationDate(_ url: URL) -> Date {
        let attributes = try? attributesOfItem(atPath: url.path)
        return attributes?[.modificationDate] as? Date ?? Date(timeIntervalSince1970: 0)
    }
}
